lib/model/stream_topics.dart

```diff
@@ -0,0 +1,220 @@
+
+  AutocompleteDataCache autocompleteDataCache = AutocompleteDataCache();
+
+  void registerMentionAutocomplete(TopicAutocompleteView view) {
+    final added = _mentionAutocompleteViews.add(view);
+    assert(added);
+  }
+
+  void unregisterMentionAutocomplete(TopicAutocompleteView view) {
+    final removed = _mentionAutocompleteViews.remove(view);
+    assert(removed);
+  }
+
+  /// Called when the app is reassembled during debugging, e.g. for hot reload.
+  ///
+  /// Calls [MentionAutocompleteView.reassemble] for all that are registered.
+  ///
+  void reassemble() {
+    for (final view in _mentionAutocompleteViews) {
+      view.reassemble();
+    }
+  }
+}
+
+/// A view-model for the recent-DM-conversations UI.
+///
+/// This maintains the list of recent DM conversations,
+/// plus additional data in order to efficiently maintain the list.
+class TopicAutocompleteView extends ChangeNotifier {
+  TopicAutocompleteView._({
+    required this.topics,
+    required this.store,
+    required this.narrow,
+  });
+
+  List<Topic>? topics;
+  final PerAccountStore store;
+  final StreamNarrow narrow;
+
+  static Future<TopicAutocompleteView> init({
+      required PerAccountStore store,
+      required StreamNarrow narrow
+  }) async {
+    final view = TopicAutocompleteView._(
+      store: store,
+      narrow: narrow,
+      topics: null,
+    );
+    store.topicAutocompleteViewManager.registerMentionAutocomplete(view);
+    return view;
+  }
+
+  @override
+  void dispose() {
+    store.topicAutocompleteViewManager.unregisterMentionAutocomplete(this);
+    super.dispose();
+  }
+
+  Future<void> fetch() async {
+    final result = await getStreamTopics(store.connection, streamId: narrow.streamId);
+    topics = result.topics;
+    notifyListeners();
+  }
+
+
+  TopicAutocompleteQuery? get query => _query;
+  TopicAutocompleteQuery? _query;
+  set query(TopicAutocompleteQuery? query) {
+    _query = query;
+    if (query != null) {
+      _startSearch(query);
+    }
+  }
+
+  /// Called when the app is reassembled during debugging, e.g. for hot reload.
+  ///
+  /// This will redo the search from scratch for the current query, if any.
+  void reassemble() {
+    if (_query != null) {
+      _startSearch(_query!);
+    }
+  }
+
+
+  Iterable<Topic> get results => _results;
+  List<Topic> _results = [];
+
+  // List<Topic> get filteredTopics => ;
+
+  _startSearch(TopicAutocompleteQuery query) async {
+    List<Topic>? newResults;
+
+    while (true) {
+      try {
+        newResults = await _computeResults(query);
+        break;
+      } on ConcurrentModificationError {
+        // Retry
+        // TODO backoff?
+      }
+    }
+
+    if (newResults == null) {
+      // Query was old; new search is in progress. Or, no listeners to notify.
+      return;
+    }
+
+    _results = newResults;
+    notifyListeners();
+  }
+
+  Future<List<Topic>?> _computeResults(TopicAutocompleteQuery query) async {
```

Comment:
This method, and its caller `_startSearch`, and the whole mechanism of `query` and `results`, is fairly tricky. It's also quite generic to the idea of autocomplete in general, not particularly specific to doing autocomplete on @-mentions vs. topics vs. anything else.

So it'd be good to share this code. To do that, a good commit structure would be:
* First, a prep commit factors it out into something generic which `MentionAutocompleteView` uses.
* Then, the PR's main commit adds `TopicAutocompleteView` as a new class that also uses the same generic mechanism.

Probably the way to structure that code-sharing is as a shared base class: something like
```dart
abstract class AutocompleteView<Q, R> extends ChangeNotifier {
  // …
  Q? get query => _query;
  // …
  Iterable<R> get results => _results;
  // …
```

I guess abstracting out `_computeResults` is slightly trickier. But even if that part stays duplicated and it's only the other parts that become shared, that'll be helpful.